index.bs

```diff
@@ -3967,6 +3987,33 @@ Note: The {{UserVerificationRequirement}} enumeration is deliberately not refere
         interest of minimizing disruption to the user interaction flow).
 </div>
 
+### Client Capabilities Set (dictionary <dfn dictionary>ClientCapabilities</dfn>) ### {#dictionary-client-capabilities}
+
+<xmp class="idl">
+    dictionary ClientCapabilities {
```

Comment:
Encoding the possible features as IDL defeats the point. We might as well have static functions in that case!

Reply:
> Encoding the possible features as IDL defeats the point. We might as well have static functions in that case!

@emlun / @MasterKale while the dictionary model is a bit cleaner, using an enum provides a stronger long term solution as previously defined here: https://w3c.github.io/webauthn/#sct-domstring-backwards-compatibility

I think this should be reverted back to an enum w/ DOMStrings.

Reply:
I'm not entirely sure how it'd map to IDL, but I do think we should define the return value in a way that would allow for devs to destructure the return value and easily know which capabilities are available (with `undefined` communicating that the client doesn't know how to determine support for that capability):

For example

```ts
type ClientCapabilities = {
  userVerifyingPlatformAuthenticator?: boolean; // true | false | undefined
  hybridTransport?: boolean;
  passkeyPlatformAuthenticator?: boolean;
  conditionalMediation?: boolean;
}

type getClientCapabilities = () => Promise<ClientCapabilities>;
```

And then in the real world devs could do this:

```js
const {
  userVerifyingPlatformAuthenticator,
  hybridTransport,
  passkeyPlatformAuthenticator,
  conditionalMediation,
} = await getClientCapabilities();
```

I think that'd make for a pleasantly usable method.

Reply:
I still think it seems significant to communicate a difference between a capability being positively unsupported vs. unknown, though. How should we do that, then? I can imagine a few options:

 1. Return two lists: features supported and features unsupported:
    ```js
    { capable: ["hybrid-transport", "conditional-mediation"], incapable: ["user-verifying-platform-authenticator"] }
    ```
 2. Return two lists: features supported and list of all known features:
    ```js
    { capable: ["conditional-mediation"], all: ["conditional-mediation", "hybrid-transport", "passkey-platform-authenticator", "user-verifying-platform-authenticator"] }
    ```
 3. Return a [`maplike<DOMString, boolean>`](https://webidl.spec.whatwg.org/#idl-maplike) instead of a `dictionary`
 4. Do none of the above, and instead forbid future additions of `ClientCapability` values that are already supported in browsers.

Of these I think I like (3) best. Thoughts, other ideas? Or should we let it be ambiguous whether a capability is unsupported or unknown?

Reply:
@emlun The problem of getting TS to understand when new capabilities drop before the official TypeScript types get updated isn't so bad. For example, in your example we could define a type that extends the official one and coerce TS into understanding:

```ts
// Platform library (e.g., Node) providing TypeScript type declaration converted from WebIDL
export type ClientCapabilities = {
  hybridTransport: boolean;
}

export type ClientCapabilitiesFuture = ClientCapabilities & {
  conditionalMediation: boolean;
};

// Application code below
const foo = await PublicKeyCredential.getClientCapabilities() as ClientCapabilitiesFuture;

// Assume browsers now support a `conditionalMediation` property but the above definition library hasn't yet been updated
console.log(foo.conditionalMediation);
```

Then projects could, on a case-by-case basis integrate additional capabilities that browsers might independently start adding return values for. But could we define a data structure in IDL that would define the four capabilities we'd include in the spec, but open the door for "but also there might be some other values in there based on browsers expanding this method's insights"?

> 3\. Return a [`maplike<DOMString, boolean>`](https://webidl.spec.whatwg.org/#idl-maplike) instead of a `dictionary`:
>     ```webidl
>     interface ClientCapabilities {
>       maplike<DOMString, boolean>;
>     }
>     ```

If we can't have the hybrid I proposed above, I suppose I'd be fine with this. The downside is that every RP using the method and TypeScript would have to independently define a "`KnownClientCapabilities`" type to case the return value to with the known values. And it'd be harder to catch misspellings...this is a tough one 🤔

Reply:
I don't know enough IDL to know when to use `maplike` or `record`, but `Promise<record<...>>` exists in Web APIs today, but `Promise<maplike<...>>` does not. (I think if we're returning an object, we can define an interface that is `maplike`, and with additional properties, or we can just record a `record` object.)

Reply:
I think this stream of comments is resolved?

Reply:
Looking at this again, I think @agl is right that `record` is probably more appropriate for this. But I'd say we could handle that in a separate PR if you want to land this one @timcappalli.